Implement a few more rvalue translation to smir

celinval opened the pull request:
Add the implementation for a few more RValue variants. For now, I simplified the stable version of `RValue::Ref` by removing the notion of Region.

r? @oli-obk

celinval committed:
Implement a few more rvalue translation to smir

 - Introduce an Opaque type for adding information that is still
   internal to the compiler.

oli-obk commented on compiler/rustc_smir/src/rustc_internal/mod.rs:
```diff
@@ -10,7 +10,7 @@ use crate::{
 pub use rustc_span::def_id::{CrateNum, DefId};
 
-fn with_tables<R>(mut f: impl FnMut(&mut Tables<'_>) -> R) -> R {
+pub(crate) fn with_tables<R>(mut f: impl FnMut(&mut Tables<'_>) -> R) -> R {
```
I wanted to keep this private for now, to have a stronger separation. Not sure yet whether that is good long term, but maybe let's keep it for now?

celinval replied:
Ok, do you have any suggestion on how to get the context when translating from an internal rustc type that requires context information?

From what I can tell, we need to modify the Tables when translation from MIR -> SMIR and we need to read the Tables when translating SMIR -> MIR. So I was thinking that this function should be available for the `rustc_smir` and the `rustc_internal` to do their respective translations.

celinval replied:
@oli-obk?

oli-obk replied:
The wrapper function you removed was my attempt at keeping everything but the wrappers in that module. As long as that isn't too annoying, I think we should keep adding such wrapper functions as needed. We can always remove all of them and expose the tables function directly, but that should be its own PR with its own motivation.

celinval replied:
That makes sense. Let me revert that. Thanks

oli-obk commented:
> by removing the notion of Region.

one thing I've done for cases like this is to just use a `String` as the "stable" version, so the data is at least there in some opaque, but printable, manner.

r=me after splitting off the Cargo.toml and extern crate changes into their own commit (and maybe the `Stable` cleanups that do not add new behaviour, too?)

@bors delegate+

bors commented:
:v: @celinval, you can now approve this pull request!

If @oli-obk told you to "`r=me`" after making some further change, please make that change, then do `@bors r=@oli-obk`
<!-- homu: {"type":"Delegated","delegator":"oli-obk","delegate":"celinval"} -->

celinval commented:
> > by removing the notion of Region.
> 
> one thing I've done for cases like this is to just use a `String` as the "stable" version, so the data is at least there in some opaque, but printable, manner.

That makes sense. What if I create an actual Opaque type that implements ToString (or any other name you might prefer). Just to differentiate from cases where String is the actual stable version.

> r=me after splitting off the Cargo.toml and extern crate changes into their own commit (and maybe the `Stable` cleanups that do not add new behaviour, too?)

Ok. Just to clarify. Do you want me to split them into separate commit or separate PRs?

> @bors delegate+

oli-obk commented:
> What if I create an actual Opaque type that implements ToString (or any other name you might prefer).

Oh heh, that's even better, as then you have an actual type that can even be used in a `Stable` impls in a way that won't require changing anything but the type and the stable impls in the future when this gets turned into a real conversion.



> Do you want me to split them into separate commit or separate PRs?

Just separate commits. It's totally alright to piggyback some cleanup commits into a regular PR

celinval committed:
Fix standalone build

Add extern declarations and optional dependencies to fix build done
directly via `cargo build`.

celinval committed:
Implement Stable for AssertMessage

bors commented:
:pushpin: Commit 844dc5ba839ed48ca6c8c82da87b49d240d51f77 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 844dc5ba839ed48ca6c8c82da87b49d240d51f77 -->
<!-- homu: {"type":"Approved","sha":"844dc5ba839ed48ca6c8c82da87b49d240d51f77","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

oli-obk commented:
bors silently ignores approvals that happen before (force-)pushes

@bors r+

bors commented:
:pushpin: Commit 97e609cf3ad37b1c95a53035ee7cd28352123b8f has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 97e609cf3ad37b1c95a53035ee7cd28352123b8f -->
<!-- homu: {"type":"Approved","sha":"97e609cf3ad37b1c95a53035ee7cd28352123b8f","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

oli-obk commented:
@bors rollup

matthiaskrgr commented:
@bors r- 
might need rebase or something
https://github.com/rust-lang/rust/actions/runs/5358354083/jobs/9720522386#step:25:1244

bors commented:
:umbrella: The latest upstream changes (presumably #113330) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

celinval commented:
@oli-obk I had to solve some conflicts. Do I need another approval?

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 633c0222d0c9296c6a7dbb57e99a44ce8088f667 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 633c0222d0c9296c6a7dbb57e99a44ce8088f667 -->
<!-- homu: {"type":"Approved","sha":"633c0222d0c9296c6a7dbb57e99a44ce8088f667","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
may need rebase

bors commented:
:umbrella: The latest upstream changes (presumably #113484) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

celinval commented:
@oli-obk sorry, another conflict...

oli-obk commented:
@bors r+ rollup

No other smir stuff in the queue, should work this time 😆

bors commented:
:pushpin: Commit b9f378b19bd5ac234139663e9df1fe8ddca5c120 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk b9f378b19bd5ac234139663e9df1fe8ddca5c120 -->
<!-- homu: {"type":"Approved","sha":"b9f378b19bd5ac234139663e9df1fe8ddca5c120","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors rollup=iffy might break clippy tests, lets find out

oli-obk commented:
@bors rollup

It cannot possibly affect clippy. No one uses the SMIR crate.
